docs/sequence.rst

```diff
@@ -0,0 +1,647 @@
+
+    case [1, 2, 3] of
+      [] -> "empty list"
+      (x:xs) -> "first:" ++ show x ++ " rest:" ++ show xs
+    > "first:1 rest:[2,3]"
+
+
+We can do the same thing with sequences, but first we'll need to enable the
+``PatternSynonyms`` language extension. At the top of your Haskell source file
+above the ``module`` declaration add::
+
+   {-# LANGUAGE PatternSynonyms #-}
+
+Now that we have that out of the way, lets pattern match on a sequence!
+
+::
+
+    case Seq.fromList [1, 2, 3] of
+      Empty -> "empty sequence"
+      x :<| xs -> "first:" ++ x ++ " rest:" ++ show xs
+    > "first:1 rest:fromList [2,3]"
+
+.. NOTE:: You can't copy/paste this into GHCi because its multiple lines.
+
+You can also take an element off the end::
+
+    case Seq.fromList [1, 2, 3] of
+      Empty -> "empty sequence"
+      xs :|> x -> "last element:" ++ x
+    > "last element:3"
+
+Querying
+^^^^^^^^
+
+Check if a sequence is empty
+""""""""""""""""""""""""""""
+
+::
+
+    Seq.null :: Seq a -> Bool
+    Seq.null xs = ...
+
+:seq:`null` returns ``True`` if the sequence ``xs`` is empty, ``False``
+otherwise.
+
+::
+
+    Seq.null Seq.empty
+    > True
+
+    Seq.null (Seq.fromList [1, 2, 3])
+    > False
+
+The length/size of a sequence
+"""""""""""""""""""""""""""""
+
+::
+
+    Seq.length :: Seq a -> Int
+    Seq.length xs = ...
+
+:seq:`length` returns the length of the sequence ``xs``.
+
+::
+
+    Seq.length Seq.empty
+    > 0
+
+    Seq.size (Seq.fromList [1, 2, 3])
+    > 3
+
+The element at a given index
+""""""""""""""""""""""""""""
+
+::
+
+    Seq.lookup :: Int -> Seq a -> Maybe a
+    Seq.lookup n xs = ...
+
+    Seq.!? :: Seq a -> Int -> Maybe a
+    xs !? n = ...
+
+:seq:`lookup` returns the element at the position ``n``, ``Nothing`` if the
+index is out of bounds. `!?
+<https://hackage.haskell.org/package/containers-0.5.10.2/docs/Data-Sequence.html#v:-33--63->`_
+is simply a flipped version of ``lookup``.
+
+.. NOTE::
+   You may need to import ``!?`` qualified if you're using a ``Map`` in the same
+   file because they both export the operator.
+
+::
+
+    Seq.index :: Seq a -> Int -> a
+    Seq.index xs n = ...
+
+:seq:`index` returns the element at the given position, throws a runtime
+``error`` if the index is out of bounds.
+
+.. TIP::
+   Use ``lookup``/``!?`` whenever you can and explicitly deal with the
+   ``Nothing`` case.
+
+::
+
+    (Seq.fromList ["base", "containers"]) Seq.!? 0
+    > Just "base"
+
+    Seq.index 0 (Seq.fromList ["base", "containers"])
+    > "base"
+
+    (Seq.fromList ["base", "containers"]) Seq.!? 2
+    > Nothing
+
+    Seq.index (Seq.fromList ["base", "containers"]) 2
+    > "*** Exception: index out of bounds
+
+
+Modification
+^^^^^^^^^^^^
+
+Inserting an element
+""""""""""""""""""""
+
+::
+
+    Seq.insertAt :: Int -> a -> Seq a -> Seq a
+    Seq.insertAt i x xs = ...
+
+:seq:`insertAt` inserts ``x`` into ``xs`` at the index ``i``, shifting the rest
+of the sequence over.
```

Comment:
Note that if the index is out of range, the element will be added at the beginning or end of the sequence as appropriate. `insertAt (-10) 'a' (fromList "hi") = fromList "ahi"`, and `insertAt 10 'p' (fromList "hi") = fromList "hip"`.

Reply:
Added words and examples demonstrating that.